src/detect-engine-build.c

```diff
@@ -1125,9 +1126,376 @@ static int RuleSetWhitelist(Signature *s)
     return wl;
 }
 
-int CreateGroupedPortList(DetectEngineCtx *de_ctx, DetectPort *port_list, DetectPort **newhead,
-        uint32_t unique_groups, int (*CompareFunc)(DetectPort *, DetectPort *));
-int CreateGroupedPortListCmpCnt(DetectPort *a, DetectPort *b);
+typedef enum {
+    RULE_WHITELISTED = 0,
+    RULE_MPM_FAST_PATTERN,
+    RULE_MPM_NEGATED,
+    RULE_NO_MPM,
+    RULE_SYN_ONLY,
+    RULE_NO_GROUP,
+} WhitelistGroupType;
+
+#define RULE_GROUP_MAX 6
+
+static int CreateGroupedPortListCmpCnt(const DetectPort *a, const DetectPort *b)
+{
+    if (a->sh->init->sig_cnt >= b->sh->init->sig_cnt) {
+        SCLogDebug("%u:%u (cnt %u, wl %d) wins against %u:%u (cnt %u, wl %d)", a->port, a->port2,
+                a->sh->init->sig_cnt, a->sh->init->score, b->port, b->port2, b->sh->init->sig_cnt,
+                b->sh->init->score);
+        return 1;
+    }
+    return -1;
+}
+
+static inline void SortGroupList(
+        DetectPort **list, int (*CompareFunc)(const DetectPort *, const DetectPort *))
+{
+    DetectPort head_ph;
+    head_ph.next = *list;
+    DetectPort *cur = *list;
+
+    while (cur != NULL && cur->next != NULL) {
+        if (CompareFunc(cur, cur->next) > 0) {
+            cur = cur->next;
+        } else {
+            DetectPort *tmp = cur->next;
+            cur->next = tmp->next;
+
+            DetectPort *insert = &head_ph;
+            while (insert != cur && CompareFunc(insert->next, tmp) > 0) {
+                insert = insert->next;
+            }
+            tmp->next = insert->next;
+            insert->next = tmp;
+
+            if (insert == cur) {
+                cur = tmp;
+            }
+        }
+    }
```

Comment:
Could use a better sorting algorithm than insertion sort. This is quadratic complexity.

Reply:
Done in https://github.com/OISF/suricata/pull/10497 by Victor